Support needsOutputs for RecordFunction and ObserverUtil improvements

louisfeng opened the pull request:
Summary:
Added needsOutputs support to RecordFunction, improved ObserverUtil functions to handle list data. Minor refactor names to be consistent.

To get output data from kernel calls, we need to temporarily capture them before passing them to the record function. Then the results are released to function return. We handle two cases, for unboxed and boxed kernels. The boxed version is fairly simple since all outputs are stored in the stack object. For unboxed kernel calls, we added a `ReturnValue` utility class to properly handle the different return values of unboxed kernels.

For optimization, this intermediate capture is only enabled for observers that request `needsOutputs(true)` and should not affect other observers or when the observer is not enabled.

Test Plan:
```
=> buck build //caffe2/test/cpp/jit: --show-output
=> buck-out/gen/caffe2/test/cpp/jit/jit --gtest_filter=RecordFunctionTest*
CUDA not available. Disabling CUDA and MultiCUDA tests
Note: Google Test filter = RecordFunctionTest*-*_CUDA:*_MultiCUDA
[==========] Running 7 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 7 tests from RecordFunctionTest
[ RUN      ] RecordFunctionTest.TracedTestInputsOutputs
[       OK ] RecordFunctionTest.TracedTestInputsOutputs (226 ms)
[ RUN      ] RecordFunctionTest.SampledCallbacks
[       OK ] RecordFunctionTest.SampledCallbacks (771 ms)
[ RUN      ] RecordFunctionTest.RecordFunctionGuard
[       OK ] RecordFunctionTest.RecordFunctionGuard (0 ms)
[ RUN      ] RecordFunctionTest.Callbacks
[       OK ] RecordFunctionTest.Callbacks (2 ms)
[ RUN      ] RecordFunctionTest.ShouldRun
[       OK ] RecordFunctionTest.ShouldRun (0 ms)
[ RUN      ] RecordFunctionTest.Basic
[       OK ] RecordFunctionTest.Basic (1 ms)
[ RUN      ] RecordFunctionTest.OperatorNameOverload
[       OK ] RecordFunctionTest.OperatorNameOverload (1 ms)
[----------] 7 tests from RecordFunctionTest (1001 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (1002 ms total)
[  PASSED  ] 7 tests.

```

Differential Revision: D25966661

louisfeng commented:
@ilia-cher, please added reviewers for Dispatcher changes. Thanks.

ilia-cher commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -403,6 +403,119 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+
+// ReturnValue is intended to capture return values from Dispatcher unboxed
+// kernel calls.
+// A record function may requests to get outputs from the kernel calls. For
```
typo

ilia-cher commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -403,6 +403,119 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+// need to handle different kinds of return values, cache them temporarily, then
+// release the values for the actual function call return.
+template <typename Return>
+struct ReturnValue {
```
@smessmer @swolchok Sebastian, Scott - any ideas about adding this template?

Can we just reuse the existing code for boxing in boxing/impl/boxing.h ?

louisfeng replied:
Thanks, trying a few ideas, will follow up.

ilia-cher commented on aten/src/ATen/record_function.cpp:
```diff
@@ -452,6 +460,8 @@ void RecordFunction::before(
   state_->sequence_nr_ = sequence_nr;
   state_->thread_id_ = currentThreadId();
   state_->operator_name_ = op.operator_name();
+  state_->op_input_size = op.schema().arguments().size();
+  state_->op_output_size = op.schema().returns().size();
```
should we just use state_'s inputs and outputs size instead in num_inputs and num_outputs?

louisfeng replied:
For boxed calls with stack, I have encountered more values on the stack than actually pushed/popped by the operator. If it works like an actual call stack, we may have values in the stack from previous calls based on this [note](https://github.com/pytorch/pytorch/wiki/Boxing-and-Unboxing-in-the-PyTorch-Operator-Library). Maybe @smessmer could chime in.

smessmer replied:
In a boxed call, the stack can contain args from calls "further up the stack", not only from the current call.

ilia-cher reviewed (commented):
LG, some questions to @smessmer @swolchok about adding ReturnValue

swolchok commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -403,6 +403,119 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+
+// tuple<Ts> specialization.
+template <typename... Ts>
+struct ReturnValue<std::tuple<Ts...>> {
```
Think we should just use return_to_ivalue and push_outputs from make_boxed_from_unboxed_functor; would definitely solve this case

smessmer replied:
I agree with @ilia-cher and @swolchok here. We already have this template machinery and should reuse it instead of building it anew. We should be able to remove all instances of the `ReturnValue` class and just our existing implementations.

louisfeng replied:
I have made some changes to use `push_outputs::copy`. This makes it explicit that we are not moving the outputs.

swolchok commented on aten/src/ATen/record_function.h:
```diff
@@ -126,11 +126,36 @@ struct TORCH_API RecordFunction {
   }
 
-  const std::vector<c10::IValue>& inputs() const {
+  inline const std::vector<c10::IValue>& inputs() const {
```
nit: putting `inline`  on a function that's already in the class body does nothing

louisfeng replied:
Yes. @ilia-cher, if you don't mind, I'll remove the other `inline` for all the member functions in this header to be consistent.

swolchok commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -820,43 +848,71 @@ std::unique_ptr<at::ObserverContext> tracedInputsCallback(
+      if (output.isTensor()) {
+        sizes.push_back(output.toTensor().sizes().vec());
+      } else if (output.isScalar()) {
+        sizes.push_back(std::vector<int64_t>());
```
nit: `sizes.emplace_back()` is shorter

louisfeng replied:
Fixed.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -403,6 +403,119 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+  // afterward. Copy elision and RVO do not apply to class data members. Using
+  // move semantic to avoid copies when possible.
+  inline Return release() {
+    return std::move(val);
```
since you're moving from the state, this should be an `&&` overload and only callable on rvalue references, i.e.

```
Return release() && {
    return std::move(val);
}
```

louisfeng replied:
We do actually use `ReturnValue` object as a lvalue. `release()` follows a similar semantic as `std::unique_ptr::release()`.

louisfeng replied:
Thanks, I have made rvalue reference required for this method to enforce correctness.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -423,6 +536,11 @@ inline Return Dispatcher::callWithDispatchKeySlowPath(const TypedOperatorHandle<
         runRecordFunction(guard, op, dispatchKey);
       }
+      if (C10_UNLIKELY(guard.needsOutputs())) {
+        detail::ReturnValue<Return> result(kernel, op, dispatchKeySet, std::forward<Args>(args)...);
```
This calls from an unboxed context, but this can call into either a boxed or an unboxed kernel. If you call into a boxed kernel, then the kernel will unbox the results for you before returning. And you'd be re-boxing them here. This is additional effort that we might be able to avoid. @ilia-cher did something similar to avoid the double boxing for the arguments.

louisfeng replied:
I have discussed with @ilia-cher. The new implementation with `push_outputs::copy` should handle this correctly.

smessmer commented on aten/src/ATen/record_function.h:
```diff
@@ -262,8 +295,11 @@ struct TORCH_API RecordFunction {
     StringView name_;
     int64_t sequence_nr_ = -1;
     std::vector<c10::IValue> inputs_;
+    std::vector<c10::IValue> outputs_;
```
if they're not always present, then it might make sense to explicitly model that using a `c10::optional`

louisfeng replied:
The vector is empty is a meaningful check for inputs or outputs. If a callback needs inputs or outputs, it expect 0 or more data. Using `c10::optional`introduces a new possible outcome (`c10::optional::has_value()`) to check and the meaning of `optional::has_value() -> false` and `vector::empty()` seem redundant.

smessmer replied:
The current implementation represents two things in the same way
  - needsOutput==true with an side-effect-only op returning zero returns
  - needsOutput==false with any op
Explicit representations with `c10::optional` are usually better and have a lower chance of somebody writing code that uses the `outputs_` variable forgetting that an empty vector can mean two different things.

louisfeng replied:
Yes, good points. Here it is being consistent with `inputs_` and existing interface.

louisfeng commented:
@smessmer @swolchok @ilia-cher, thanks for the feedbacks. I have addressed most of the minor changes. I will look into using existing boxing utilities to capture return values.

louisfeng commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -403,6 +403,119 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+            std::forward<Args>(args)...)} {}
+  // Wrap the return values in a vector of IValues. We also specialize for
+  // containers below, so they get
+  inline std::vector<IValue> getOutputs() {
```
Fixed.

louisfeng commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -403,6 +403,119 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+inline std::vector<IValue> ReturnValue<std::vector<at::Tensor>>::getOutputs() {
+  std::vector<IValue> outputs;
+  for (auto& t : val) {
+    outputs.emplace_back(IValue(t));
```
Thanks, I'll double check.

louisfeng replied:
I found in `buck-out/dev/gen/caffe2/aten/gen_aten=core/TensorMethods.cpp/core/TensorMethods.cpp`, we have a few instances where the return values are `std::vector<Tensor>`:
```
// aten::unsafe_chunk(Tensor self, int chunks, int dim=0) -> Tensor[]
std::vector<Tensor> Tensor::unsafe_chunk(int64_t chunks, int64_t dim) const {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::unsafe_chunk", "")
        .typed<std::vector<Tensor> (const Tensor &, int64_t, int64_t)>();
    return op.call(const_cast<Tensor&>(*this), chunks, dim);
}

// aten::chunk(Tensor(a) self, int chunks, int dim=0) -> Tensor(a)[]
std::vector<Tensor> Tensor::chunk(int64_t chunks, int64_t dim) const {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::chunk", "")
        .typed<std::vector<Tensor> (const Tensor &, int64_t, int64_t)>();
    return op.call(const_cast<Tensor&>(*this), chunks, dim);
}

// aten::tensor_split.sections(Tensor(a) self, int sections, int dim=0) -> Tensor(a)[]
std::vector<Tensor> Tensor::tensor_split(int64_t sections, int64_t dim) const {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::tensor_split", "sections")
        .typed<std::vector<Tensor> (const Tensor &, int64_t, int64_t)>();
    return op.call(const_cast<Tensor&>(*this), sections, dim);
}
```

louisfeng replied:
But in this case, we probably want to handle it as a single return ivalue vs. tuple:
```
// aten::cummax(Tensor self, int dim) -> (Tensor values, Tensor indices)
std::tuple<Tensor,Tensor> Tensor::cummax(int64_t dim) const {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::cummax", "")
        .typed<std::tuple<Tensor,Tensor> (const Tensor &, int64_t)>();
    return op.call(const_cast<Tensor&>(*this), dim);
}

// aten::cummax.dimname(Tensor self, Dimname dim) -> (Tensor values, Tensor indices)
std::tuple<Tensor,Tensor> Tensor::cummax(Dimname dim) const {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::cummax", "dimname")
        .typed<std::tuple<Tensor,Tensor> (const Tensor &, Dimname)>();
    return op.call(const_cast<Tensor&>(*this), dim);
}
```

louisfeng replied:
The correct behavior is `vector<Tensor>` is treated as a single return value. Where as tuple<Tensor> is treated as multiple return values. Fixed.

louisfeng commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -403,6 +403,119 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+  // containers below, so they get
+  inline std::vector<IValue> getOutputs() {
+    std::vector<IValue> outputs;
+    outputs.emplace_back(IValue(val));
```
Thanks, will fix `emplace_back()`. The reason we don't use `move` here is because we actually want to return `val` later.

louisfeng replied:
This is fixed.

louisfeng commented:
@swolchok @smessmer, I took a look at push_outputs, it's pretty close to what I need. But can't seem to use them as is. Some comments on the issues: 
* The `ReturnType` for the kernel call could be `Tensor&`, in which case, I can't decay it similar to make_boxed_from_unboxed_functor unless we change the interface `inline Return Dispatcher::callWithDispatchKeySlowPath` to also decay the Return type. This is not a problem, similar to `struct return_to_ivalue<at::Tensor&, AllowDeprecatedTypes, void>`, which also does this specialization. I can keep the specialized `inline at::Tensor& ReturnValue<at::Tensor&>::release()` method.
* To use `push_outputs`, it expects the output to be rvalue reference and internally tries to move output. Since ultimately we still want to release the output to for `callWithDispatchKeySlowPath` to return, we probably don't want `push_outputs` to do the move, i.e. why I probably don't want to use `std::move(output_)` in `impl::push_outputs<ReturnType, false>::call(output_, &stack);`, but this will result in compiler error as expected.

swolchok commented:
> To use push_outputs, it expects the output to be rvalue reference and internally tries to move output. Since ultimately we still want to release the output to for callWithDispatchKeySlowPath to return, we probably don't want push_outputs to do the move, i.e. why I probably don't want to use std::move(output_) in impl::push_outputs<ReturnType, false>::call(output_, &stack);, but this will result in compiler error as expected.

If `ReturnType` is not a reference, you can do `impl::push_outputs<ReturnType, false>::call(ReturnType(output_, &stack))` to explicitly give push_outputs a fresh copy to move from.

louisfeng commented:
> > To use push_outputs, it expects the output to be rvalue reference and internally tries to move output. Since ultimately we still want to release the output to for callWithDispatchKeySlowPath to return, we probably don't want push_outputs to do the move, i.e. why I probably don't want to use std::move(output_) in impl::push_outputs<ReturnType, false>::call(output_, &stack);, but this will result in compiler error as expected.
> 
> If `ReturnType` is not a reference, you can do `impl::push_outputs<ReturnType, false>::call(ReturnType(output_, &stack))` to explicitly give push_outputs a fresh copy to move from.

Thanks, one issue is if the `ReturnType` is a `vector<T>`, this copy could be expensive. In this case, we are trying to pass this value to get wrapped by `IValue` in `return_to_ivalue` which should be cheap by keeping a pointer to the object. So I added a `copy()` method to `push_outputs` to handle this use case. Please review and see if this makes sense. Thanks!

louisfeng commented:
@swolchok @smessmer @ilia-cher, I have made changes to address most of the feedbacks, and replied with comments. Please take a look and see if they make sense, thanks.

smessmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -477,23 +484,35 @@ namespace impl {
       torch::jit::push(*stack, return_to_ivalue<OutputType, AllowDeprecatedTypes>::call(std::forward<OutputType>(output)));
     }
+    static void copy(const OutputType& output, Stack* stack) {
+      torch::jit::push(*stack, return_to_ivalue<OutputType, AllowDeprecatedTypes>::copy(std::forward<const OutputType>(output)));
```
don't think you need the `std::forward` here since you're copying anyways

louisfeng replied:
Fixed, thanks.

smessmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -477,23 +484,35 @@ namespace impl {
     }
+    template<size_t... indices>
+    static void copy_(const std::tuple<OutputTypes...>& output, Stack* stack, std::index_sequence<indices...>) {
+      torch::jit::push(*stack, return_to_ivalue<OutputTypes, AllowDeprecatedTypes>::copy(std::forward<const OutputTypes>(std::get<indices>(output)))...);
```
same here

louisfeng replied:
Fixed, thanks.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -402,9 +402,70 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+
+// Handle the lvalue reference differently since it should not be moved.
+template <>
+inline at::Tensor& ReturnValue<at::Tensor&>::release() && {
```
This shouldn't be needed. The generic implementation above

```
ReturnType relese() && {
  return std::move(output_);
}
``

will not move lvalues. Calling `std::move` doesn't actually move, it only creates an rvalue reference to the object. And if `ReturnValue` is an lvalue reference, then that rvalue reference gets converted back into an lvalue reference on return and no harm done.

louisfeng replied:
Without this specialization, we'd get a compile error: 
```
In file included from buck-out/dev/gen/caffe2/aten/gen_aten=core/TensorMethods.cpp/core/TensorMethods.cpp:9:
caffe2/aten/src/ATen/core/dispatch/Dispatcher.h:437:12: error: non-const lvalue reference to type 'at::Tensor' cannot bind to a temporary of type 'typename std::remove_reference<Tensor &>::type' (aka 'at::Tensor')
    return std::move(output_);
```

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -402,9 +402,70 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+struct ReturnValue {
+  template <typename F, typename... Args>
+  ReturnValue(
+      F&& kernel,
```
if you're not forwarding the kernel anywhere, then `const F& kernel` works and is easier to understand than `F&& kernel`. Same for the other specialization below

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -402,9 +402,70 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
+  ReturnValue(
+      F&& kernel,
+      const TypedOperatorHandle<ReturnType(Args...)>& op,
+      DispatchKeySet& dispatchKeySet,
```
any reason the dispatchKeySet here needs to be mutable?

louisfeng replied:
Fixed.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -423,6 +484,15 @@ inline Return Dispatcher::callWithDispatchKeySlowPath(const TypedOperatorHandle<
+      if (C10_UNLIKELY(guard.needsOutputs())) {
+        // Call the kernel and capture the outputs temporarily in result to pass
+        // to RecordFunction.
+        detail::ReturnValue<Return> result(
```
I find it confusing that the constructor of `ReturnValue` does the actual call to the kernel. If I was scanning this op to try and find the actual kernel call, that's not where I would have expected it. It might be better to do the kernel call in here and then just pass the `Return` object into the `ReturnValue` and wrap/store it there.

louisfeng replied:
Yeah, this is a bit tricky. Here are some cases to consider:

- `Return -> void`: this case we have to do some special handling, one method is to use `constexpr` similar to `make_boxed_from_unboxed_functor::call`, or we could do something like `ReturnValue` with template specialization. Wish we have C++17, without it, the next one is tricky. 
- `Return -> obj&`: With `constexpr` without C++17, we'll need to deal with `guts::if_constexpr<has_outputs>` lambda. How do we get it out of the lambda call? Can we do it without a copy? Not very familiar with `guts::if_constexpr`. The simpler way seems to initialize `output_` in constructor, hence calling the kernel in constructor.

So I have been thinking about doing this a few different ways, but each has some trade offs to balance perf and usability. There might be some cool ideas I haven't thought about.

louisfeng commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -377,6 +377,10 @@ namespace impl {
     }
+    static IValue copy(const T& v) {
+      assert_is_valid_output_type<T, AllowDeprecatedTypes>();
+      return IValue(v);
```
I suppose we could use `c10::ivalue::from(v);`, and it will likely not actually move `v` due to `const` (but can't prevent a ctor that matches `(const value&&)`). Explicitly construct here seems clearer.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -486,6 +556,10 @@ inline void Dispatcher::callBoxed(const OperatorHandle& op, Stack* stack) const
     kernel.callBoxed(op, dispatchKeySet, stack);
+    // track outputs
+    if (C10_UNLIKELY(guard.isActive() && guard.needsOutputs())) {
+      guard.setOutputs(*stack);
```
This copies the **whole** stack, not only the outputs for the current operator. Is this intentional? Perf might be better if you only copy the outputs for the current operator.

louisfeng replied:
This follows the same approach as `needsInputs()` see `runRecordFunction(guard, op, dispatchKey, *stack);` at line 551. This is to be consistent with existing implementation. If we change the existing interface, there might be other code impacted, so we should get some more feedbacks. Maybe another diff to address this.

ilia-cher replied:
ok, we can follow up as part of a separate perf investigation and improvement pr

ilia-cher replied:
using `last(stack, num_outputs)` to get the last `op.schema().returns().size()` items would be better, but ok let's do it in a follow up to address this and inputs too

louisfeng committed:
Support needsOutputs for RecordFunction and ObserverUtil improvements (#54442)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/54442

Added needsOutputs support to RecordFunction, improved ObserverUtil functions to handle list data. Minor refactor names to be consistent.

To get output data from kernel calls, we need to temporarily capture them before passing them to the record function. Then the results are released to function return. We handle two cases, for unboxed and boxed kernels. The boxed version is fairly simple since all outputs are stored in the stack object. For unboxed kernel calls, we added a `ReturnValue` utility class to properly handle the different return values of unboxed kernels.

For optimization, this intermediate capture is only enabled for observers that request `needsOutputs(true)` and should not affect other observers or when the observer is not enabled.

Test Plan:
```
=> buck build //caffe2/test/cpp/jit: --show-output
=> buck-out/gen/caffe2/test/cpp/jit/jit --gtest_filter=RecordFunctionTest*
CUDA not available. Disabling CUDA and MultiCUDA tests
Note: Google Test filter = RecordFunctionTest*-*_CUDA:*_MultiCUDA
[==========] Running 7 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 7 tests from RecordFunctionTest
[ RUN      ] RecordFunctionTest.TracedTestInputsOutputs
[       OK ] RecordFunctionTest.TracedTestInputsOutputs (226 ms)
[ RUN      ] RecordFunctionTest.SampledCallbacks
[       OK ] RecordFunctionTest.SampledCallbacks (771 ms)
[ RUN      ] RecordFunctionTest.RecordFunctionGuard
[       OK ] RecordFunctionTest.RecordFunctionGuard (0 ms)
[ RUN      ] RecordFunctionTest.Callbacks
[       OK ] RecordFunctionTest.Callbacks (2 ms)
[ RUN      ] RecordFunctionTest.ShouldRun
[       OK ] RecordFunctionTest.ShouldRun (0 ms)
[ RUN      ] RecordFunctionTest.Basic
[       OK ] RecordFunctionTest.Basic (1 ms)
[ RUN      ] RecordFunctionTest.OperatorNameOverload
[       OK ] RecordFunctionTest.OperatorNameOverload (1 ms)
[----------] 7 tests from RecordFunctionTest (1001 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (1002 ms total)
[  PASSED  ] 7 tests.

```

Differential Revision: D25966661

fbshipit-source-id: 69f91973ef9c52425ad1025fb93d11a82da747e9

louisfeng commented:
@smessmer @swolchok @ilia-cher: I have made a few more changes to address feedbacks and improve clarity. There are some good discussions involve further interface changes that impact existing code. I recommend we address them in a separate PR once we better understand their impacts on the dependent code.

swolchok reviewed (commented):
seems OK to me, leaving for Sebastian & Ilia

louisfeng commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -486,6 +558,10 @@ inline void Dispatcher::callBoxed(const OperatorHandle& op, Stack* stack) const
     // keeping the guard alive while executing the kernel
     kernel.callBoxed(op, dispatchKeySet, stack);
+    // track outputs
+    if (C10_UNLIKELY(guard.isActive() && guard.needsOutputs())) {
```
Fixed, thanks.
